nulogy/design-system
# NDS-229 improves tests, build and start scripts

# Problem

Although the project is still pretty small it is obviously that as it grows and the dependency graph becomes more complex, it will be a chore to figure out how to build and test packages along with their (downstream) dependencies and (upstream) dependents.

# What I did

1. Introduced [Lerna](https://lernajs.io) to make managing and running scripts across packages easier.
1. Improves scripts in the root pakcage.json. See [Script docs](https://deploy-preview-31--nulogy-design-system.netlify.com/guides/scripts) in Catalog for details, but briefly:
    - **`yarn test`** runs all tests in all packages. Uses lerna to build only the downstream dependencies that have changed on this branch before running the tests. *There is no watch mode for tests yet*, but I have created a ticket (NDS-233) to add a watch mode.
    - **Start scripts**: are now just responsible for starting a single dev server. It's expected that developers start a build process as well (see the `watch` command below). There is also a `yarn all` to start and sanity check the whole project.
    - **Build scripts**:  `build` builds all, `build:public` only builds the library packages (css, components and tokens) and `watch` builds library packages in watch mode. Thanks to Lerna all of these scripts are able to build in dependency-graph order.
    - **Clean scripts** to quickly remove generated artifacts. `yarn clean` to remove just the builds,  `yarn pristine` returns the repo to a freshly checked out state.
1. Updates [Guides / Package Scripts](https://deploy-preview-31--nulogy-design-system.netlify.com/guides/scripts) to document the above changes.
1. Simplified the @nulgy/components package.json by removing unused configuration and dependencies introduced by CRA.
1. Simplifies and unifies jest configuration in components and tokens packages (the only packages with tests currently).
1. Updates Netlify build settings to run new lerna-based `build` command.

# To test

1. run `yarn pristine` to remove all `dist/` and `node_modules/`.
1. run `yarn install` to reinstall dependencies.
1. run `yarn test` to run all tests.
1. [Check the docs]( https://deploy-preview-31--nulogy-design-system.netlify.com/guides/scripts):
    1. Are they clear? 
    1. Any typos?
    1. Can you use them to establish a comfortable workflow to develop eg. components or css? Etc.
